api_paper_2024/example_3/similarity_docking.py

```diff
@@ -0,0 +1,505 @@
+        )
+
+        self.start_point_writer = EntryWriter(
+            os.path.join(self.settings.output_directory, 'start_points.mol2')
+        )
+
+    def run(self):
+        """Runs the job."""
+        self.get_arguments()
+        self.session = self.docker.dock(
+            os.path.join(self.settings.output_directory, 'settings.conf'),
+            'interactive'
+        )
+        self.run_substrate()
+        self.done = set()
+        dock_index = 1
+        cycles = 0
+        while cycles < self.args.maximum_cycles:
+            print(f"Start cycle {cycles}")
+            hits = self.search()
+            if not hits:
+                break
+
+            cycles += 1
+            for h in hits:
+                self.done.add(h.identifier)
+                mol = h.molecule
+
+                if mol.all_atoms_have_sites is False:
+                    print('%-8s: No 3d structure' % h.identifier)
+                    self.fail_writer.write(mol)
+                    continue
+                mol.assign_bond_types()
+                mol.add_hydrogens()
+
+                # Skip this one as its beyond the weight threshold
+                if self.args.maximum_weight and mol.molecular_weight > self.args.maximum_weight:
+                    continue
+
+                entry = Entry.from_molecule(mol)
+                poses = self.session.dock(entry)
+                dock_index += 1
+                if not poses:
+                    print('%-8s: Failed to dock' % h.identifier)
+                    continue
+                score, pose = self.best(poses)
+                self.write_poses(poses=poses, dock_index=dock_index)
+                self.write_scores(score=score, hit=h, pose=pose)
+
+    def search(self):
+        # If we have too many start points, chop off the lowest scoring ones
+        self.best_poses = sorted(self.best_poses, key=lambda x: x[0])  # Only sort on
+        if self.args.maximum_start_points is not None and len(self.best_poses) > self.args.maximum_start_points:
+            self.best_poses = self.best_poses[len(self.best_poses) - self.args.maximum_start_points:]
+
+        if self.args.search_chembl:
+            return self.search_chembl()
+        else:
+            return self.search_csd()
+
+    def search_chembl(self):
+        try:
+            return self._search_chembl()
+        except ChemblSearchError as e:
+            print(f"Chembl Searching failed with an exception {e} - script will terminate")
+            return None
+
+    def search_csd(self):
+        try:
+            return self._search_csd()
+        except RuntimeError as e:
+            print(f"CSD Searching failed with an exception {e} - script will terminate")
+            return None
+
+    def get_chembl_molecule(self, chemblid):
+        """ Get a SMILEs entry from its ID
+
+        :param chemblid:
+        :return:
+        """
+
+        molecule = new_client.molecule
```

Comment:
The Chembl implementation seems to be missing from this.

Reply:
The user has to install it via PIP

Reply:
The point is earlier in the piece, if the user has asked on the commandline to import the optional CHEMbl dependencies it will appear. (I.e. the code has called import_chembl_dependencies() )

The point is by doing it this way we can fail early but only log spurious import failures if the user wants to use CHEMbl.

So IMHO reviewdog is being a *bad dog*